Fixing errors introduced by recent JAX release

DanPuzzuoli opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add it in the CHANGELOG file under Unreleased section.
⚠️ If your pull request fixes an open issue, please link to the issue.

✅ I have added the tests to cover my changes.
✅ I have updated the documentation accordingly.
✅ I have read the CONTRIBUTING document.
-->

### Summary

Needed to add additional `Array` wrapping in one spot in `rotating_wave_approximation` in which a raw JAX array was controlling a binary operation.

Also, in `final_state_converter`, JAX arrays were being fed into quantum info types, e.g. `DensityMatrix`, raising an error. I've changed this function to always convert the array to a `numpy.array` before wrapping in `cls`, as `cls` will always be a quantum info type, which can only accept `numpy.array`s in any case. The doc string for this function has been updated to explain this.

chriseclectic reviewed (commented):
If this is going into a patch release there should be a release note for `fixes` saying what the bug that was fixed is. I think in this case it is something like 

> Fixes a bug where simulations could raise an error using JAX solvers for Statevector and DensityMatrix final states.

DanPuzzuoli commented:
Updates:
- Added release note, as suggested.
- Discovered that new JAX release advanced sparse support enough that LindbladModel.evaluate_rhs can now be reverse mode differentiated when in sparse mode. Removed this caveat from the documentation, and changed the test case for this function to test reverse-mode autodiff instead of forward mode (reverse is more restrictive so it's enough to just test this one). This test change was done by deleting the forward mode test case, which overrode an inherited reverse-mode test case.
- Due to the above, bumped the minimum JAX version, and explained this in the release note.
